### PR TITLE
Update KV Cache to use num_kv_heads instead of num_heads

### PR DESCRIPTION
**Pull request by mirceamironenco**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Update KV Cache to use num_kv_heads instead of num_heads for more memory-efficient generation.
* Simplified/adapted mha fwd pass to have the expand happen after the kv cache.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

If one changes `generation.yaml` to use llama 3.1 8b (num_kv_heads < num_heads), and the prompt/max_new_toks:
```
INFO:torchtune.utils._logging:Running InferenceRecipe with resolved config:

checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: /tmp/Meta-Llama-3.1-8B-Instruct/
  checkpoint_files:
  - model-00001-of-00004.safetensors
  - model-00002-of-00004.safetensors
  - model-00003-of-00004.safetensors
  - model-00004-of-00004.safetensors
  model_type: LLAMA3
  output_dir: /tmp/Meta-Llama-3.1-8B-Instruct/
  recipe_checkpoint: null
device: cuda
dtype: bf16
enable_kv_cache: true
max_new_tokens: 4096
model:
  _component_: torchtune.models.llama3_1.llama3_1_8b
prompt:
  system: null
  user: Tell me a long joke.
quantizer: null
seed: 1234
temperature: 0.6
tokenizer:
  _component_: torchtune.models.llama3.llama3_tokenizer
  max_seq_len: null
  path: /tmp/Meta-Llama-3.1-8B-Instruct/original/tokenizer.model
  prompt_template: null
top_k: 300
```

I get with the new kv cache (on RTX A6000):
```
INFO:torchtune.utils._logging:Time for inference: 12.72 sec total, 22.79 tokens/sec
INFO:torchtune.utils._logging:Bandwidth achieved: 379.93 GB/s
INFO:torchtune.utils._logging:Memory used: 17.01 GB
```
Old kv cache:
```
INFO:torchtune.utils._logging:Time for inference: 11.86 sec total, 24.46 tokens/sec
INFO:torchtune.utils._logging:Bandwidth achieved: 447.16 GB/s
INFO:torchtune.utils._logging:Memory used: 18.62 GB
```
I'm not sure how relevant the tok/sec change is given that this is not compiled and batch_size=1.

### CODE REVIEW

File: torchtune/modules/kv_cache.py [hunk 1]
```diff
@@ -53,7 +51,7 @@ def reset(self) -> None:
 
     @property
     def size(self) -> int:
-        return self.cache_pos[0].item()
+        return int(self.cache_pos[0].item())
```
Comment (SalmanMohammadi) [inline torchtune/modules/kv_cache.py:54]: how come this was added?

Comment (mirceamironenco) [reply torchtune/modules/kv_cache.py:54]: It's a leftover from a version that would raise a type error. I'll remove it, thanks!

File: torchtune/modules/attention.py [hunk 2]
```diff
@@ -258,47 +258,36 @@ def forward(
...
+            # k,v shape: [b, s_y, n_kv, h_d]
             k = k.view(b, s_y, -1, self.head_dim)
+            v = v.view(b, s_y, -1, self.head_dim)
             if self.pos_embeddings is not None:
                 k = self.pos_embeddings(k, input_pos=input_pos)
 
-            # View + expand + reshape bring num_kv_heads to num_heads for k and v
-            # to match q.
+            # k,v shape: [b, n_kv, s_y, h_d]
+            k, v = k.transpose(1, 2), v.transpose(1, 2)
```
Comment (SalmanMohammadi) [inline torchtune/modules/attention.py:274]: mega nit:
```suggestion
            k = k.transpose(1, 2)
            v = v.transpose(1, 2)
```
helps keeps everything nice and readble

File: torchtune/modules/attention.py [hunk 3]
```diff
@@ -258,47 +258,36 @@ def forward(
...
-                v = v.expand(b, s_y, self.num_kv_heads, q_per_kv, self.head_dim)
-
-            # [b, s, n_h, h_d]
-            k = k.reshape(b, s_y, -1, self.head_dim)
-            v = v.reshape(b, s_y, -1, self.head_dim)
-
-            # [b, n_h, s, h_d]
-            k = k.transpose(1, 2)
-            v = v.transpose(1, 2)
+                expand_shape = (-1, -1, q_per_kv, -1, -1)
```
Comment (SalmanMohammadi) [inline torchtune/modules/attention.py:283]: I like this

File: torchtune/modules/attention_utils.py [hunk 4]
```diff
@@ -183,7 +183,6 @@ def _attention_call(
             dropout_p: float,
             is_causal: bool,
         ) -> torch.Tensor:
-
```
Comment (SalmanMohammadi) [inline torchtune/modules/attention_utils.py:186]: Could remove this : )
